### ISSUE TITLE
SMP: preemptionPoint() should know about other cores

### ISSUE THREAD

**Issue by Indanz** [seL4/seL4#1054]
To limit WCET, `preemptionPoint()` is called to check if there is anything else to do. It returns `EXCEPTION_PREEMPTED` when IRQs are pending or when the current task runs out of budget.

But with SMP, the global kernel lock will be held and other cpu cores can be blocked on the lock. `preemptionPoint()` either needs to check for other cores wanting to take the lock, or to always return `EXCEPTION_PREEMPTED` on SMP. 

This issue was raised on the [mailing list](https://lists.sel4.systems/hyperkitty/list/devel@sel4.systems/thread/SLX75KFHL2MONFNLUFWOKJT6J25I4ISF/) by Yadong Li.

**Comment by lsf37** [comment]
To clarify the `bug` label: there is no functional correctness problem, but it certainly is a kernel WCET problem and it is inconsistent with the intention of `preemptionPoint()`.

The user-level effect is that threads on other cores may be blocked from entering the kernel for the entirety of a long-running delete operation instead of being able to make progress. This is less worse than in the unicore case, where all threads would be blocked from making any progress, but it is still not ideal.

**Comment by Indanz** [comment]
> This is less worse than in the unicore case, where all threads would be blocked from making any progress, but it is still not ideal.

On unicore only threads with lower priority are blocked. On multicore _all_ threads and interrupts will be blocked on all other cores. So it's much worse there, that's why it's a bug.

**Comment by lsf37** [comment]
> On unicore only threads with lower priority are blocked. 

On unicore nothing else runs, because we are in the kernel and won't get out of the kernel until `preemptionPoint()` decides that it is Ok to preempt. There is no bug on unicore, but if there were, the impact would be much worse.

> On multicore _all_ threads and interrupts will be blocked on all other cores. So it's much worse there, that's why it's a bug.

They will only be blocked when they want to enter the kernel. As long as they don't make a kernel call, they complete their time slice, which would usually be much longer than even a long running delete operation.

But not being able to get into the kernel absolutely is bad enough, I fully agree that it needs to be fixed.
